src/server.js

```diff
@@ -0,0 +1,245 @@
+        ws.terminate()
+        log_peer('terminated')
+      }
+
+      ws.on('pong', () => {
+        ws.alive = true
+      })
+      ws.on('error', error => {
+        log_peer(error)
+        ws.force_terminate()
+      })
+      ws.on('close', () => {
+        ws.force_terminate()
+      })
+
+      pipeline(
+          on(ws, 'message'),
+          async function *(source) {
+            for await (const chunk of source) {
+              const {
+                id: operation_id, document, variables, unsub,
+              } = JSON.parse(chunk.toString())
+
+              if (unsub) {
+                const stream = user_streams.get(operation_id)
+                if (stream) {
+                  log_peer('stream %s was terminated', operation_id)
+                  stream.end()
+                }
+
+                continue
+              }
+
+              yield {
+                id: operation_id,
+                document,
+                variables,
+              }
+            }
+          },
+          executor.generate.bind(executor),
+          async source => {
+            for await (const {
+              id: stream_id, stream,
+            } of source) {
+              const existing_stream = user_streams.get(stream_id)
+              if (existing_stream) existing_stream.end()
+              user_streams.set(stream_id, stream)
+
+              pipeline(
+                  stream,
+                  async stream_source => {
+                    for await (const chunk of stream_source) {
+                      if (!ws.alive) {
+                        ws.force_terminate()
+                        return
+                      }
+
+                      const {
+                        id: query_id,
+                        operation_type,
+                        operation_name,
+                        data,
+                        errors,
+                      } = chunk
+                      log_peer(
+                          'sending %O[%O]: %O',
+                          operation_type,
+                          operation_name,
+                    errors?.length ? errors : data,
+                      )
+                      ws.send(JSON.stringify({
+                        id: query_id,
+                        operation_type,
+                        operation_name,
+                        data,
+                        errors,
+                      }))
+                    }
+                  },
+                  error => {
+                    if (error) console.error(error)
+                    else log_peer('operation terminated')
+                  },
+              )
+            }
+          },
+          () => {
+            log_peer('disconnected')
+          },
+      )
+    })
+
+    const interval = setInterval(() => {
+      wss.clients.forEach(ws => {
+        if (!ws.alive) {
+          ws.force_terminate()
+          return
+        }
+
+        ws.alive = false
```

Comment:
You realy need to handle this yourself ?

Reply:
seems so https://github.com/websockets/ws#how-to-detect-and-close-broken-connections